src/modules/resource/drain.c

```diff
@@ -557,6 +558,66 @@ static void undrain_cb (flux_t *h,
     idset_destroy (idset);
 }
 
+/* Add rank to ids, adjusting rank if the rank:host mapping has changed.
+ * Don't add the rank if the host no longer exists, or if it exceeds
+ * the instance size.
+ *
+ * N.B. When running multiple brokers per node, flux_get_rankbyhost()
+ * returns the first rank on 'host', so its result cannot be directly
+ * used as the new rank.  Instead, first check that flux_get_hostbyrank()
+ * differs from 'host'.
+ */
+static void add_target (struct idset *ids,
+                        unsigned int rank,
+                        const char *host,
+                        flux_t *h)
+{
+    if (host) {
+        const char *nhost = flux_get_hostbyrank (h, rank);
```

Comment:
Is there any case where the `hostlist` attribute would not be available at this time?

I think booting from config the hostlist is built from `[bootstrap]`, but in the case where the broker is booting from PMI, do all ranks have to be online if the hostlist is built from PMI? Perhaps this is only a theoretical problem, but I'm thinking of the case of restarting a batch instance. This probably isn't even possible now, but the idea of remapping ranks based on hostnames dynamically made me think this is a step perhaps towards restarting a batch instance on a different set of nodes.

(Sorry for the tangent there)

Reply:
> do all ranks have to be online if the hostlist is built from PMI

Yes, they all have to be online so this shouldn't be an issue now.

This is going to be a problem though.  We currently have a built in assumption that the `hostlist` attribute is set during bootstrap, before rc1 starts, and that it never changes after that (it is immutable and therefore gets cached on first use in clients).  I'm not sure how to resolve that but you're right when we do this may need to change.

The same problem exists with R reranking.

More thought required.